libevm/triedb/firewood/firewood.go

```diff
@@ -0,0 +1,518 @@
+// Thus, we recognize the single root C as the only proposal, and dereference it.
+func (*Database) Dereference(common.Hash) {}
+
+// Firewood does not support this.
+func (*Database) Cap(common.StorageSize) error {
+	return nil
+}
+
+func (db *Database) Close() error {
+	db.proposalLock.Lock()
+	defer db.proposalLock.Unlock()
+
+	// All remaining proposals must be dereferenced.
+	db.possibleProposals = nil
+	db.proposalMap = nil
+	db.proposalTree = nil
+
+	// Close the database
+	// We must provide a context since it may hang while waiting for the finalizers to complete.
+	return db.Firewood.Close(context.Background())
+}
+
+func (db *Database) Update(root, parent common.Hash, block uint64, _ *trienode.MergedNodeSet, _ *triestate.Set, opts ...stateconf.TrieDBUpdateOption) error {
+	// We require block hashes to be provided for all blocks in production.
+	// However, many tests cannot reasonably provide a block hash for genesis, so we allow it to be omitted.
+	parentHash, hash, ok := stateconf.ExtractTrieDBUpdatePayload(opts...)
+	if !ok {
+		log.Error("firewood: no block hash provided for block %d", block)
+	}
+
+	// The rest of the operations except key-value arranging must occur with a lock
+	db.proposalLock.Lock()
+	defer db.proposalLock.Unlock()
+
+	// Check if this proposal already exists.
+	// During reorgs, we may have already created this proposal.
+	// Additionally, we may have already created this proposal with a different block hash.
+	if existingProposals, ok := db.proposalMap[root]; ok {
+		for _, existing := range existingProposals {
+			// If the block hash is already tracked, we can skip proposing this again.
+			if _, exists := existing.info.hashes[hash]; exists {
+				log.Debug("firewood: proposal already exists", "root", root.Hex(), "parent", parent.Hex(), "block", block, "hash", hash.Hex())
+				return nil
+			}
+			// We already have this proposal, but should create a new context with the correct hash.
+			// This solves the case of a unique block hash, but the same underlying proposal.
+			if _, exists := existing.info.parent.hashes[parentHash]; exists {
+				log.Debug("firewood: proposal already exists, updating hash", "root", root.Hex(), "parent", parent.Hex(), "block", block, "hash", hash.Hex())
+				existing.info.hashes[hash] = struct{}{}
+				return nil
+			}
+		}
+	}
+
+	// We must use one of the unverified proposals if it exists.
+	var h *proposal
+	for _, possible := range db.possibleProposals {
+		if _, ok := possible.info.parent.hashes[parentHash]; ok {
+			h = possible
+			break
+		}
+	}
+	if h == nil {
+		// Edge case: first set of proposals before `Commit`, or empty genesis block
+		// Neither `Update` nor `Commit` is called for genesis, so we can accept a proposal with parentHash of empty.
+		for _, possible := range db.possibleProposals {
+			if possible.info.parent.hashes[common.Hash{}] == struct{}{} {
```

Comment:
Since `hashes` is `map[common.Hash]struct{}`, accessing a non-existent key returns the zero value `struct{}{}`, making this comparison always true when the key doesn't exist. This is a logic bug - it should check if the key exists.

```suggestion
            if _, ok := possible.info.parent.hashes[common.Hash{}]; ok {
```

Reply:
fixed